### PR TITLE
(PR-SERIES-P4) [REFACTOR]: Memory handler

### PR DESCRIPTION
**Pull request by amadolid**

### CODE REVIEW

File: jaclang/runtimelib/architype.py [hunk 1]
```diff
@@ -110,22 +160,6 @@ class NodeAnchor(Anchor):
 
     architype: Optional[NodeArchitype] = None
     edges: list[EdgeAnchor] = field(default_factory=list)
-    edge_ids: list[UUID] = field(default_factory=list)
-
-    def populate_edges(self) -> None:
```
Comment (ypkang) [inline jaclang/runtimelib/architype.py:115]: where did this go?

File: jaclang/runtimelib/architype.py [hunk 2]
```diff
@@ -14,14 +14,34 @@
...
+class AnchorState:
+    """Anchor state handler."""
+
+    # None == do nothing
+    # False == to be deleted
+    # True == already deleted
+    deleted: bool | None = None
+    connected: bool = False
+    persistent: bool = False
+    hash: int = 0
```
[the remark this answers is not in the document]
Comment (ypkang) [reply jaclang/runtimelib/architype.py:27]: One benefit is it will make the serializing ancho simpler. We won't need to specifically handle `state`

File: jaclang/runtimelib/context.py [hunk 3]
```diff
@@ -7,73 +7,51 @@
...
         jac_program = JacProgram(mod_bundle=None, bytecode=None)
         self.jac_machine.attach_program(jac_program)
 
     def init_memory(self, base_path: str = "", session: str = "") -> None:
         """Initialize memory."""
-        if session:
-            self.mem = ShelveStorage(session)
-        else:
-            self.mem = Memory()
+        self.mem = ShelfStorage(session)
```
Comment (ypkang) [inline jaclang/runtimelib/context.py:32]: If user doesn't specify session, it should be in memory only i think.

File: jaclang/runtimelib/context.py [hunk 4]
```diff
@@ -7,73 +7,51 @@
...
+                root = Root().__jac__
+                self.mem.set(root.id, root)
+
+            if not isinstance(root.architype, Root):
                 raise ValueError(f"Invalid root object: {root}")
             else:
                 self.root = root.architype
         return self.root
 
-    def get_obj(self, obj_id: UUID) -> Anchor | None:
```
Comment (ypkang) [inline jaclang/runtimelib/context.py:58]: I think an abstraction at the context level might be useful. For example, if the memory interface changes later, we just need to update this one place, without having to update throughout the codebase.

File: jaclang/runtimelib/memory.py [hunk 5]
```diff
@@ -2,98 +2,148 @@
...
+            for id in ids
+            if (anchor := self.__mem__.get(id)) and (not filter or filter(anchor))
+        )
 
-    def has_obj_in_store(self, obj_id: UUID) -> bool:
-        """Check if the object exists in the underlying store."""
-        return obj_id in self.mem
+    def find_one(
+        self,
+        ids: ID | Iterable[ID],
```
Comment (ypkang) [inline jaclang/runtimelib/memory.py:45]: for `find_one`, should this just be a single ID?

File: jaclang/runtimelib/memory.py [hunk 6]
```diff
@@ -2,98 +2,148 @@
...
+        """Remove anchor/s from memory."""
+        if not isinstance(ids, Iterable):
+            ids = [ids]
 
-    def close(self) -> None:
-        """Close any connection, if applicable."""
-        self.mem.clear()
+        for id in ids:
+            self.__mem__.pop(id, None)
+            self.__gc__.add(id)
```
Comment (ypkang) [inline jaclang/runtimelib/memory.py:62]: What is the usage of the `__gc__` list in this scenario?

File: jaclang/runtimelib/memory.py [hunk 7]
```diff
@@ -2,98 +2,148 @@
...
-    def connect(self, session: str) -> None:
-        """Connect to storage."""
-        self.session = session
-        self.storage = shelve.open(session)
+        self.__shelf__ = open(session) if session else None  # noqa: SIM115
 
     def close(self) -> None:
-        """Close the storage."""
+        """Close memory handler."""
+        if isinstance(self.__shelf__, Shelf):
```
Comment (ypkang) [inline jaclang/runtimelib/memory.py:78]: I think we can just do `if self.__shelf__` here. It is easier to read. 

Checking for type makes reader think this can be other types.

File: jaclang/runtimelib/memory.py [hunk 8]
```diff
@@ -2,98 +2,148 @@
...
+                        and not isinstance(d.architype, Root)
+                        and not d.edges
+                    ):
+                        d.state.deleted = True
+                        self.__shelf__.pop(_id, None)
+                    elif d.state.hash != (new_hash := hash(dumps(d))):
+                        d.state.hash = new_hash
+                        self.__shelf__[_id] = d
+
+            self.__shelf__.sync()
```
Comment (ypkang) [inline jaclang/runtimelib/memory.py:114]: Need to walk through this part.

**Review by ypkang — CHANGES_REQUESTED** [review]

File: jaclang/runtimelib/architype.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (amadolid) [reply jaclang/runtimelib/architype.py:27]: there's a reason why I add every mutable on a different class inside Anchors
we can discuss later

File: jaclang/runtimelib/context.py (hunk 4 above)
Comment (amadolid) [reply jaclang/runtimelib/context.py:58]: my assumptions is, we will also have

(not exact*)
Jac.get_context()
Jac.get_datasource()

I think ExecutionContext should be holder only (or atleast initializer)

File: jaclang/runtimelib/memory.py (hunk 5 above)
Comment (amadolid) [reply jaclang/runtimelib/memory.py:45]: it can be multiple id, but find first

File: jaclang/runtimelib/memory.py (hunk 6 above)
Comment (amadolid) [reply jaclang/runtimelib/memory.py:62]: to be deleted anchors

File: jaclang/runtimelib/memory.py (hunk 7 above)
Comment (amadolid) [reply jaclang/runtimelib/memory.py:78]: make sense. Will update

File: jaclang/plugin/default.py [hunk 9]
```diff
@@ -500,26 +502,26 @@ def disconnect(
...
+                    and (src_arch := source.sync())
+                    and (trg_arch := target.sync())
                 ):
                     if (
                         dir in [EdgeDir.OUT, EdgeDir.ANY]
                         and node == source
                         and trg_arch in right
                     ):
-                        anchor.detach()
+                        anchor.destroy()
```
[the remark this answers is not in the document]
Comment (amadolid) [reply jaclang/plugin/default.py:517]: detach is essentially destroying the edge

destroy calls detach

File: jaclang/cli/cli.py [hunk 10]
```diff
@@ -124,8 +124,8 @@ def run(
     if not node or node == "root":
         entrypoint: Architype = Jac.get_root()
     else:
-        obj = Jac.context().get_obj(UUID(node))
-        if obj is None or obj.architype is None:
+        obj = Jac.context().mem.find_one(UUID(node))
```
[the remark this answers is not in the document]
Comment (amadolid) [reply jaclang/cli/cli.py:127]: it should be Jac.datasource or Jac.get_datasource and this will be included on next PR

Execution should only holds "Context" related functions/attr

File: jaclang/plugin/default.py [hunk 11]
```diff
@@ -500,26 +502,26 @@ def disconnect(
             node = i.__jac__
             for anchor in set(node.edges):
                 if (
-                    (architype := anchor.architype)
+                    (architype := anchor.sync(node))
```
[the remark this answers is not in the document]
Comment (amadolid) [reply jaclang/plugin/default.py:505]: yup, we can discuss

but in Context, it's very similar in concept how ORM works

File: jaclang/runtimelib/memory.py (hunk 7 above)
Comment (amadolid) [reply jaclang/runtimelib/memory.py:78]: ooow forgot about this.

It's necessary as empty self.__shelf__ will tagged as false similar to dict

Comment (amadolid) [reply jaclang/runtimelib/memory.py:78]: ooow forgot about this.

It's necessary as empty self.\_\_shelf__ will be tagged as false similar to dict

File: jaclang/runtimelib/architype.py (hunk 1 above)
Comment (amadolid) [reply jaclang/runtimelib/architype.py:115]: we don't need this anymore as populate edge also get it one by one
not in bulk

using sync() approach will also do it like this but on demand

File: jaclang/runtimelib/context.py (hunk 3 above)
Comment (amadolid) [reply jaclang/runtimelib/context.py:32]: ShelfStorage is still a Memory if you don't specify a session

File: jaclang/cli/cli.py [hunk 12]
```diff
@@ -154,7 +154,7 @@ def get_object(id: str, session: str = "") -> dict:
     else:
         id_uuid = UUID(id)
 
-    obj = Jac.context().get_obj(id_uuid)
+    obj = Jac.context().mem.find_one(id_uuid)
     if obj is None:
         print(f"Object with id {id} not found.")
```
Comment (marsninja) [inline jaclang/cli/cli.py:159]: use logging instead of print here

File: jaclang/runtimelib/memory.py (hunk 5 above)
Comment (ypkang) [reply jaclang/runtimelib/memory.py:45]: We will have find and find_by_id which just accept a single id with no filte.r

File: jaclang/runtimelib/memory.py (hunk 8 above)
Comment (ypkang) [reply jaclang/runtimelib/memory.py:114]: This will change.

File: jaclang/runtimelib/architype.py (hunk 1 above)
Comment (ypkang) [reply jaclang/runtimelib/architype.py:115]: This will be back.

File: jaclang/runtimelib/architype.py [hunk 13]
```diff
@@ -110,22 +168,6 @@ class NodeAnchor(Anchor):
 
     architype: Optional[NodeArchitype] = None
     edges: list[EdgeAnchor] = field(default_factory=list)
-    edge_ids: list[UUID] = field(default_factory=list)
```
Comment (ypkang) [inline jaclang/runtimelib/architype.py:113]: Bring this back for now.

**Review by ypkang — CHANGES_REQUESTED** [review]

File: jaclang/runtimelib/architype.py [hunk 14]
```diff
@@ -14,14 +14,42 @@
...
+@dataclass
+class Report:
+    """Report Handler."""
+
+    id: str
+    context: dict[str, Any]
+
+
+@dataclass
+class AnchorState:
```
Comment (ypkang) [inline jaclang/runtimelib/architype.py:26]: AnchorState will no longer be neccessary.

**Comment by amadolid** [comment]
backup: https://github.com/Jaseci-Labs/jaclang/tree/pr-series/4-backup

File: jaclang/cli/cli.py (hunk 12 above)
Comment (ypkang) [reply jaclang/cli/cli.py:159]: @amadolid

File: jaclang/runtimelib/architype.py [hunk 15]
```diff
@@ -59,44 +57,38 @@ def unlinked_architype(self) -> Architype | None:
...
                 if is_dataclass(self.architype) and not isinstance(self.architype, type)
                 else {}
             ),
-        }
+        )
 
     def __hash__(self) -> int:
         """Override hash for anchor."""
-        return hash(self.ref_id)
+        return hash(self.id)
```
Comment (ypkang) [inline jaclang/runtimelib/architype.py:85]: should we override the hash here? later on, we are calling hash(dump(anchor)). If we don't override it here then we can just call hash(anchor) right?

File: jaclang/runtimelib/architype.py [hunk 16]
```diff
@@ -231,17 +222,25 @@ def spawn_call(self, walk: WalkerAnchor) -> WalkerArchitype:
...
-            state.update(
-                {
-                    "edges": [],
-                    "edge_ids": self.edge_ids or [edge.id for edge in self.edges],
-                }
-            )
+        state.update(
+            {
+                "edges": [],
+                "edge_ids": self.edge_ids + [edge.id for edge in self.edges],
```
Comment (ypkang) [inline jaclang/runtimelib/architype.py:241]: during getstate, edge_ids should always be empty right? Is there a scenario where edge_ids would have value?

File: jaclang/runtimelib/architype.py (hunk 15 above)
Comment (amadolid) [reply jaclang/runtimelib/architype.py:85]: as far as I remember, current anchor is not hashable and it's required to be overriden to be able to used in set or dict key

File: jaclang/runtimelib/architype.py (hunk 16 above)
Comment (amadolid) [reply jaclang/runtimelib/architype.py:241]: if you haven't trigger any edge_ref/disconnect and you triggered connect

File: jaclang/cli/cli.py (hunk 12 above)
Comment (amadolid) [reply jaclang/cli/cli.py:159]: might need to pr it seperately as most of unit test still validates in "print"

File: jaclang/runtimelib/architype.py (hunk 15 above)
Comment (amadolid) [reply jaclang/runtimelib/architype.py:85]: hash(dump(anchor)) is for different usage as it for including everything inside the anchor while __hash__ is just for hashing anchor's reference

File: jaclang/runtimelib/architype.py [hunk 17]
```diff
@@ -261,24 +254,15 @@ def populate_nodes(self) -> None:
...
-            obj = Jac.context().get_obj(self.source_id)
-            if obj is None:
-                raise ValueError(f"Node with id {self.source_id} not found.")
-            elif not isinstance(obj, NodeAnchor):
-                raise ValueError(f"Object with id {self.source_id} is not a node.")
-            else:
-                self.source = obj
-                self.source_id = None
+            self.source = jmem.find_by_id(self.source_id)
+            self.source_id = None
```
Comment (marsninja) [inline jaclang/runtimelib/architype.py:261]: Based on long discussion with @ypkang , we've converged on moving all logic detecting whether  an anchor is just an ID or should be loaded should live in the anchor class, similar to what you were doing wiht sync, but perhaps in this style (using property and setter):

image

File: jaclang/runtimelib/architype.py [hunk 18]
```diff
@@ -59,44 +57,38 @@ def unlinked_architype(self) -> Architype | None:
...
-
-        if architype := self.unlinked_architype():
-            state["architype"] = architype
-
-        return state
+        return {
+            "id": self.id,
+            "architype": self.unlinked_architype(),
+            "persistent": self.persistent,
+        }
```
[the remark this answers is not in the document]
Comment (ypkang) [reply jaclang/runtimelib/architype.py:64]: Create a new type PickableAnchor

File: jaclang/runtimelib/architype.py (hunk 17 above)
Comment (ypkang) [reply jaclang/runtimelib/architype.py:261]: load and unload

**Review by ypkang — CHANGES_REQUESTED** [review]

File: jaclang/runtimelib/architype.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (amadolid) [reply jaclang/runtimelib/architype.py]: need to discuss again as this will include unnecessary fields that's not needed if follow the sync/unsync approach

File: jaclang/runtimelib/architype.py (hunk 17 above)
Comment (amadolid) [reply jaclang/runtimelib/architype.py]: this will require every other lazy loaded field to be added as property to.

**Review by marsninja — APPROVED** [review]
